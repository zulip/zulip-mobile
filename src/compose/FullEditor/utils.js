/* @flow */
import regexValidator from './webLinkValidator';

export const replaceBetween = (text: string, selection: Object, what: string) =>
  text.substring(0, selection.start) + what + text.substring(selection.end);

export const isStringWebLink = (text: string): boolean => {
  const pattern = regexValidator;
  return pattern.test(text);
};
