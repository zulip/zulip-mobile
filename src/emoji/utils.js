/* @flow */
import type { RealmEmojiType, ZulipExtraEmojiType } from '../types';
import { codePointMap } from './emojiMap';

export const getRealmEmojiWithName = (realmEmoji: Object, emojiName: string): RealmEmojiType =>
  realmEmoji[Object.keys(realmEmoji).find(key => realmEmoji[key].name === emojiName)];

export const getZulipExtraEmojiWithName = (
  zulipExtraEmojis: Object,
  emojiName: string,
): ZulipExtraEmojiType => zulipExtraEmojis[emojiName];

export const getEmojiTypeFromName = (
  realmEmoji: Object,
  zulipExtraEmojis: Object,
  emojiName: string,
): string | null => {
  if (getZulipExtraEmojiWithName(zulipExtraEmojis, emojiName)) {
    return 'zulip_extra_emoji';
  } else if (getRealmEmojiWithName(realmEmoji, emojiName)) {
    return 'realm_emoji';
  } else if (codePointMap[emojiName]) {
    return 'unicode_emoji';
  }
  return null;
};

export const getEmojiCodeFromName = (
  realmEmoji: Object,
  zulipExtraEmojis: Object,
  emojiName: string,
): string | null => {
  if (getZulipExtraEmojiWithName(zulipExtraEmojis, emojiName)) {
    return emojiName;
  }
  const emoji = getRealmEmojiWithName(realmEmoji, emojiName);
  if (emoji) {
    return emoji.id.toString();
  } else if (codePointMap[emojiName]) {
    return codePointMap[emojiName];
  }

  return null;
};
