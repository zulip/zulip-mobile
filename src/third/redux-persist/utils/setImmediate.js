const hasNativeSupport = typeof global !== 'undefined' && typeof global.setImmediate !== 'undefined'
const setImmediate = hasNativeSupport ? (fn, ms) => global.setImmediate(fn, ms) : (fn, ms) => setTimeout(fn, ms)

export default setImmediate
