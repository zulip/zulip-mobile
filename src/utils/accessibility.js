/* @flow */
export const textWithUnreadCount = (text: string, unreadCount: number = 0): string => {
  if (unreadCount === 0) {
    return text;
  }

  return `${text}, ${unreadCount} unread message${unreadCount > 1 ? 's' : ''}`;
};
