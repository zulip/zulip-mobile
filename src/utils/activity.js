/* @flow */
import differenceInSeconds from 'date-fns/difference_in_seconds';

import type { Auth } from '../types';
import { focusPing } from '../api';

const APP_ACTIVITY_UPDATE_FREQUENCY = 60;
const USER_INPUT_ACTIVITY_UPDATE_FREQUENCY = 5;

let lastAppActivity = new Date();
let lastInputActivity = new Date();

export const registerAppActivity = (auth: Auth, hasFocus: boolean = true) => {
  if (differenceInSeconds(new Date(), lastAppActivity) > APP_ACTIVITY_UPDATE_FREQUENCY) {
    focusPing(auth, hasFocus, false);
    lastAppActivity = new Date();
  }
};

export const registerUserInputActivity = (auth: Auth) => {
  if (differenceInSeconds(new Date(), lastInputActivity) > USER_INPUT_ACTIVITY_UPDATE_FREQUENCY) {
    focusPing(auth, true, true);
    lastInputActivity = new Date();
  }
};
