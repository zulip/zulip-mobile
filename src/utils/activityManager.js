import {
  AppState,
} from 'react-native';
import throttle from 'lodash.throttle';

import { focusPing } from '../api';
import { Auth } from '../api/apiFetch';

// Send once per minute.
// Send when restored from suspended state
// Send on new compose message
// Send on new narrow event

export default class ActivityManager {

  auth: Auth;

  constructor(auth: Auth) {
    this.auth = auth;

    AppState.addEventListener('change', this.handleAppStateChange);
  }

  // componentWillUnmount() {
  //   AppState.removeEventListener('change', this.handleAppStateChange);
  // }

  appActivity() {
    throttle(() => focusPing(this.props.auth, true, false), 1000 * 60);
  }

  userInputActivity() {
    throttle(() => focusPing(this.props.auth, true, false), 1000 * 5);
  }

  handleAppStateChange = (currentAppState) => {
    if (currentAppState === 'active') {
      this.appActivity();
    }
  }
}
