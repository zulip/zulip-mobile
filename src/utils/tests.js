import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import StylesProvider from '../StylesProvider';
import store from '../store';

export const rendererWithStore = WrappedComponent =>
  renderer.create(
    <Provider store={store}>
      {WrappedComponent}
    </Provider>,
  );

export const rendererWithStyle = WrappedComponent =>
  renderer.create(
    <StylesProvider>
      {WrappedComponent}
    </StylesProvider>,
  );

export const rendererWithStoreAndStyle = WrappedComponent =>
  renderer.create(
    <Provider store={store}>
      <StylesProvider>
        {WrappedComponent}
      </StylesProvider>
    </Provider>,
  );
