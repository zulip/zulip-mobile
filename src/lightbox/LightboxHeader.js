/* @flow */
import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

import ModalNavBar from '../nav/ModalNavBar';
import { SlideAnimationView, Avatar } from '../common';
import { shortTime, humanDate } from '../utils/date';
import { PopRouteAction } from '../types';

const customStyles = StyleSheet.create({
  text: {
    paddingLeft: 10,
    paddingTop: 2,
    paddingBottom: 2,
    justifyContent: 'space-around',
  },
  name: {
    color: 'white',
    fontSize: 14,
    lineHeight: 14,
  },
  navBar: {
    backgroundColor: 'transparent',
    paddingTop: 0,
    borderBottomWidth: 0,
  },
  subheader: {
    color: 'white',
    fontSize: 12,
    lineHeight: 12,
  },
  children: {
    justifyContent: 'flex-start',
    paddingRight: 10,
  },
});

type Props = {
  popRoute: PopRouteAction,
  senderName: string,
  timestamp: number,
  from: number,
  to: number,
  movement: string,
  style: Object,
  avatarUrl: string,
  realm: string,
  styles: () => null,
};

export default ({ popRoute, senderName, timestamp, styles, ...restProps }: Props) => {
  const displayDate = humanDate(new Date(timestamp * 1000));
  const time = shortTime(new Date(timestamp * 1000));
  const subheader = `${displayDate} at ${time}`;

  return (
    <SlideAnimationView property={'translateY'} {...restProps}>
      <ModalNavBar
        itemsColor="white"
        rightItem={{ name: 'ios-close-outline', onPress: popRoute, style: { fontSize: 36 } }}
        style={customStyles.navBar}
        childrenStyle={customStyles.children}
        isRightItemNav
        popRoute
      >
        <Avatar {...restProps} />
        <View style={customStyles.text}>
          <Text style={[styles.username, customStyles.name]} numberOfLines={1}>
            {senderName}
          </Text>
          <Text style={customStyles.subheader} numberOfLines={1}>
            {subheader}
          </Text>
        </View>
      </ModalNavBar>
    </SlideAnimationView>
  );
};
