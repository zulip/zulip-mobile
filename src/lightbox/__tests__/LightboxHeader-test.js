import React from 'react';

import { Header } from '../LightboxHeader';
import { rendererWithStoreAndStyle } from '../../utils/tests';

describe('LightboxHeader', () => {
  test('all required props are passed correctly', () => {
    const message = {
      type: 'stream',
      avatar_url: 'https://example/tester.png',
      display_recipient: 'stream',
      timestamp: 1501218744,
      sender_full_name: 'tester',
    };
    const actions = {
      navigateBack: jest.fn(),
    };
    const auth = {
      realm: 'https://example.com',
    };
    const subheader = 'Today at 5:00 PM';

    const header = rendererWithStoreAndStyle(
      <Header
        onPressBack={actions.navigateBack}
        timestamp={message.timestamp}
        avatarUrl={message.avatar_url}
        senderName={message.sender_full_name}
        realm={auth.realm}
        subheader={subheader}
      />,
    );

    expect(header).toMatchSnapshot();
  });
});
