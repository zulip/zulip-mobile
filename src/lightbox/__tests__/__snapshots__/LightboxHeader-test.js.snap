// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LightboxHeader all required props are passed correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderBottomWidth": 1,
        "borderColor": "rgba(127, 127, 127, 0.25)",
        "flexDirection": "row",
        "height": 64,
        "justifyContent": "space-between",
        "paddingTop": 20,
      },
      Object {
        "backgroundColor": "transparent",
        "borderBottomWidth": 0,
        "paddingTop": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "center",
        },
        Object {
          "justifyContent": "flex-start",
          "paddingRight": 10,
        },
      ]
    }
  >
    <View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "height": 32,
              "width": 32,
            },
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View>
          <Image
            resizeMode="contain"
            source={
              Object {
                "uri": "https://example/tester.png",
              }
            }
            style={
              Object {
                "borderRadius": 4,
                "height": 32,
                "width": 32,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "space-around",
          "paddingBottom": 2,
          "paddingLeft": 10,
          "paddingTop": 2,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "flex": 1,
              "fontSize": 16,
              "fontWeight": "bold",
              "lineHeight": 16,
            },
            Object {
              "color": "white",
              "fontSize": 14,
              "lineHeight": 14,
            },
          ]
        }
      >
        tester
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "white",
            "fontSize": 12,
            "lineHeight": 12,
          }
        }
      >
        Today at 5:00 PM
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 44,
            "justifyContent": "center",
            "width": 44,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 12,
              },
              Array [
                Object {
                  "fontSize": 26,
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 36,
                },
              ],
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          ÔêÜ
        </Text>
      </View>
    </View>
  </View>
</View>
`;
